### PR TITLE
CDN - PowerShell Functionality and Doc Improvements

### PR DESCRIPTION
**Pull request by CamSoper**
Made a ton of improvements for usability and documentation.
- Modified all "Get-" cmdlets to return collections where a specific resource wasn't named.
- Modified all cmdlets to accept pipeline input objects.  E.g., `Get-AzureRmCdnProfile` now returns all profiles a user has access to. `Get-AzureRmCdnProfile | Get-AzureRmCdnEndpoint` will list all endpoints on all profiles a user has access to.  `Get-AzureRmCdnProfile | Get-AzureRmCdnEndpoint | Get-AzureRmCdnCustomDomain` will list all the custom domains.  Etc., etc. 
- Modified collection outputs to enumerate properly for Where-Object and ForEach-Object
- Cleaned up cmdlet parameter help for grammar and consistency
- Cleaned up help XML for grammar and consistency.
- Refactored anyplace that was using exceptions for flow control, since this is [generally frowned upon](http://www.hanselman.com/blog/GoodExceptionManagementRulesOfThumb.aspx).
- Annotated my commits.
- Created new tests for improved pipeline functionality
- Re-recorded all test sessions.

Still to-do: The help XML needs Examples.

@kuangweizhang

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@CamSoper__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (casoper). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

**Comment by hovsepm** [comment]
@azuresdkci add to whitelist

File: src/ResourceManager/Cdn/AzureRM.Cdn.psd1 [hunk 1]
```diff
@@ -9,7 +9,7 @@
 @{  
   
 # Version number of this module.  
-ModuleVersion = '1.0.5'
```
Comment (hovsepm) [inline src/ResourceManager/Cdn/AzureRM.Cdn.psd1]: please revert this change. It is part of our process to update versions in each release.

File: src/ResourceManager/Cdn/Commands.Cdn.Test/ScenarioTests/EndpointTests.ps1 [hunk 2]
```diff
@@ -355,4 +355,38 @@ function Test-EndpointCrudAndActionWithStoppedEndpoint
...
+function Test-EndpointPipeline
+{
+    $profileName = getAssetName
+    $resourceGroup = TestSetup-CreateResourceGroup
+    $resourceLocation = "EastUS"
+    $profileSku = "StandardVerizon"
+    $createdProfile = New-AzureRmCdnProfile -ProfileName $profileName -ResourceGroupName $resourceGroup.ResourceGroupName -Location $resourceLocation -Sku $profileSku
+
+    $endpointName1 = getAssetName
+	$endpointName2 = getAssetName
```
Comment (hovsepm) [inline src/ResourceManager/Cdn/Commands.Cdn.Test/ScenarioTests/EndpointTests.ps1]: please replace TAB symbol with 4 spaces in this file

File: src/ResourceManager/Cdn/Commands.Cdn.Test/ScenarioTests/ProfileTests.ps1 [hunk 3]
```diff
@@ -80,7 +80,40 @@ function Test-ProfileDeleteAndSsoWithPiping
...
+}
+
+<#
+.SYNOPSIS
+Profile cycle with piping multiple profiles down the pipeline
+#>
+function Test-ProfilePipeline
+{
+    $profileName1 = getAssetName
+	$profileName2 = getAssetName
```
Comment (hovsepm) [inline src/ResourceManager/Cdn/Commands.Cdn.Test/ScenarioTests/ProfileTests.ps1]: please replace TAB symbol with 4 spaces in this file

**Comment by hovsepm** [comment]
> 

revert version change.

---
Refers to: src/ResourceManager/Cdn/Commands.Cdn/Properties/AssemblyInfo.cs:30 in a669f26. [](commit_id = a669f261dfb68c81b3321e143a42aaa418b2fabb, deletion_comment = False)

**Comment by CamSoper** [comment]
Should be fixed now.  Sorry about that, @hovsepm !

**Comment by markcowl** [comment]
@CamSoper Test failures, looks like soem tests need to be re-recorded

**Comment by CamSoper** [comment]
@markcowl Light is green, trap is clean!

**Comment by CamSoper** [comment]
@kuangweizhang  - @hovsepm needs you to bless this, please.  :)

Thanks!

File: src/ResourceManager/Cdn/Commands.Cdn/CustomDomain/NewAzureRmCdnCustomDomain.cs [hunk 4]
```diff
@@ -24,74 +24,70 @@
...
-                    CustomDomainName, 
-                    EndpointName, 
-                    ProfileName,
-                    ResourceGroupName).Wait();
+                ResourceGroupName = CdnEndpoint.ResourceGroupName;
+                ProfileName = CdnEndpoint.ProfileName;
+                EndpointName = CdnEndpoint.Name;
+            }
 
+            var existingCustomDomain = CdnManagementClient.CustomDomains.ListByEndpoint(EndpointName, ProfileName, ResourceGroupName).Where(cd => cd.Name == CustomDomainName).FirstOrDefault();
```
Comment (kuangweizhang) [inline src/ResourceManager/Cdn/Commands.Cdn/CustomDomain/NewAzureRmCdnCustomDomain.cs]: I think we should ignore case when compare the custom domain name

Comment (kuangweizhang) [inline src/ResourceManager/Cdn/Commands.Cdn/CustomDomain/NewAzureRmCdnCustomDomain.cs]: Same comment for other places

File: src/ResourceManager/Cdn/Commands.Cdn/Endpoint/RemoveAzureRmCdnEndnpoint.cs [hunk 5]
```diff
@@ -54,35 +58,25 @@ public override void ExecuteCmdlet()
                 EndpointName = CdnEndpoint.Name;
             }
 
-            try
+            var existingEndpoint = CdnManagementClient.Endpoints.ListByProfile(ProfileName, ResourceGroupName).Where(e => e.Name == EndpointName).FirstOrDefault();
```
Comment (kuangweizhang) [inline src/ResourceManager/Cdn/Commands.Cdn/Endpoint/RemoveAzureRmCdnEndnpoint.cs]: ignore case

**Comment by kuangweizhang** [comment]
Just some comments on comparing the name of the resources, otherwise, LGTM

**Comment by CamSoper** [comment]
Thanks, @kuangweizhang! 

Implemented the changes as suggested.  Should be good to merge, I think.

File: src/ResourceManager/Cdn/Commands.Cdn.Test/ScenarioTests/CustomDomainTests.ps1 [hunk 6]
```diff
@@ -61,9 +63,10 @@ Gets and removes custom domain with stopped endpoint
 #>
 function Test-CustomDomainGetRemoveWithStoppedEndpoint
 {
-    $endpointName = "sdktest-cee91bb1-996a-44f1-96e6-dceaa6707def"
-    $hostName = "sdktest-3acbafc1-3f50-4fa4-9132-6d8c944d67e9.azureedge-test.net"    
-    $customDomainName = getAssetName
+    $endpointName = "sdktest-b0939e74-75ba-4558-afe6-edc5c19ea713"
```
Comment (markcowl) [inline src/ResourceManager/Cdn/Commands.Cdn.Test/ScenarioTests/CustomDomainTests.ps1]: The best practice is to make setup part of test execution, so that these values do not have to be updated whenever the test is re-recorded.  Please file an issue to do setup as part of test execution

File: src/ResourceManager/Cdn/Commands.Cdn/CustomDomain/RemoveAzureRmCdnCustomDomain.cs [hunk 7]
```diff
@@ -63,48 +67,38 @@ public override void ExecuteCmdlet()
...
+            {
+                throw new PSArgumentException(string.Format(Resources.Error_DeleteNonExistingCustomDomain,
+                    CustomDomainName,
+                    EndpointName,
+                    ProfileName,
+                    ResourceGroupName));
             }
 
-            if (!ShouldProcess(string.Format(
+           if (Force || ShouldProcess(string.Format(
```
Comment (markcowl) [inline src/ResourceManager/Cdn/Commands.Cdn/CustomDomain/RemoveAzureRmCdnCustomDomain.cs]: This is not the pattern for PowerShell Confirmation.  Force should only be used with a ShouldContinue prompt, after a ShouldProcess check.  Force is really only for specially destructive actions, you are already gestting a ShouldProcess prompt in most cases beacuase your ConfirmImpact is High.

Please remove the Force prompt and revert this conditional change

File: src/ResourceManager/Cdn/Commands.Cdn/CustomDomain/GetAzureRmCdnCustomDomain.cs [hunk 8]
```diff
@@ -18,38 +18,61 @@
 using Microsoft.Azure.Commands.Cdn.Models.CustomDomain;
 using Microsoft.Azure.Commands.Cdn.Properties;
 using Microsoft.Azure.Management.Cdn;
+using System.Linq;
+using Microsoft.Azure.Commands.Cdn.Models.Endpoint;
 
 namespace Microsoft.Azure.Commands.Cdn.CustomDomain
 {
     [Cmdlet(VerbsCommon.Get, "AzureRmCdnCustomDomain"), OutputType(typeof(PSCustomDomain))]
     public class GetAzureRmCdnCustomDomain : AzureCdnCmdletBase
```
Comment (markcowl) [inline src/ResourceManager/Cdn/Commands.Cdn/CustomDomain/GetAzureRmCdnCustomDomain.cs]: For all cmdlets where you are adding a parameter set - you need to have a DefaultParameterSetName promprty in your Cmdlet attribute

File: src/ResourceManager/Cdn/Commands.Cdn/Endpoint/NewAzureRmCdnEndpoint.cs [hunk 9]
```diff
@@ -23,111 +23,105 @@
 using Microsoft.Azure.Commands.Cdn.Properties;
 using Microsoft.Azure.Management.Cdn;
 using Microsoft.Azure.Management.Cdn.Models;
+using Microsoft.Azure.Commands.Cdn.Models.Profile;
 
 namespace Microsoft.Azure.Commands.Cdn.Endpoint
 {
     [Cmdlet(VerbsCommon.New, "AzureRmCdnEndpoint"), OutputType(typeof(PSEndpoint))]
```
Comment (markcowl) [inline src/ResourceManager/Cdn/Commands.Cdn/Endpoint/NewAzureRmCdnEndpoint.cs]: SupportsShouldProcess = true

File: src/ResourceManager/Cdn/Commands.Cdn/Endpoint/NewAzureRmCdnEndpoint.cs [hunk 10]
```diff
@@ -23,111 +23,105 @@
...
 
         [Parameter(Mandatory = false,
-            HelpMessage = "The tags to associate with the Azure Cdn Endpoint")]
+            HelpMessage = "The tags to associate with the Azure CDN endpoint.")]
         public Hashtable Tags { get; set; }
 
         public override void ExecuteCmdlet()
         {
-            try
+            if(ParameterSetName == ObjectParameterSet)
```
Comment (markcowl) [inline src/ResourceManager/Cdn/Commands.Cdn/Endpoint/NewAzureRmCdnEndpoint.cs]: Guard execution with ShouldProcess

File: src/ResourceManager/Cdn/Commands.Cdn/Endpoint/PublishAzureRmCdnEndpointContent.cs [hunk 11]
```diff
@@ -23,27 +23,27 @@ namespace Microsoft.Azure.Commands.Cdn.Endpoint
     [Cmdlet(VerbsData.Publish, "AzureRmCdnEndpointContent"), OutputType(typeof(bool))]
```
Comment (markcowl) [inline src/ResourceManager/Cdn/Commands.Cdn/Endpoint/PublishAzureRmCdnEndpointContent.cs]: SupportsShouldProcess=true

File: src/ResourceManager/Cdn/Commands.Cdn/Endpoint/PublishAzureRmCdnEndpointContent.cs [hunk 12]
```diff
@@ -23,27 +23,27 @@ namespace Microsoft.Azure.Commands.Cdn.Endpoint
...
-        [Parameter(Mandatory = true, HelpMessage = "The resource group of the Azure Cdn Profile")]
+        [Parameter(Mandatory = true, HelpMessage = "The resource group of the Azure CDN profile.")]
         [ValidateCount(Constants.PurgeLoadMinimumCollectionCount, Constants.PurgeLoadMaximumCollectionCount)]
         public string[] LoadContent { get; set; }
 
-        [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "Return object if specified.")]
+        [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "Return object (if specified).")]
         public SwitchParameter PassThru { get; set; }
 
         public override void ExecuteCmdlet()
```
Comment (markcowl) [inline src/ResourceManager/Cdn/Commands.Cdn/Endpoint/PublishAzureRmCdnEndpointContent.cs]: Guard execution with ShouldProcess()

File: src/ResourceManager/Cdn/Commands.Cdn/Endpoint/RemoveAzureRmCdnEndnpoint.cs [hunk 13]
```diff
@@ -20,31 +20,35 @@
...
-        [Parameter(Mandatory = true, ParameterSetName = ObjectParameterSet, ValueFromPipeline = true, HelpMessage = "The endpoint.")]
+        [Parameter(Mandatory = true, ParameterSetName = ObjectParameterSet, ValueFromPipeline = true, HelpMessage = "The CDN endpoint object.")]
         [ValidateNotNull]
         public PSEndpoint CdnEndpoint { get; set; }
 
-        [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "Return object if specified.")]
+        [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "Return object (if specified).")]
         public SwitchParameter PassThru { get; set; }
 
+        [Parameter()]
```
Comment (markcowl) [inline src/ResourceManager/Cdn/Commands.Cdn/Endpoint/RemoveAzureRmCdnEndnpoint.cs]: See comments on previous remove.  Force/SHouldContinue are only for extra prompting for especially destructive situations

File: src/ResourceManager/Cdn/Commands.Cdn/CustomDomain/RemoveAzureRmCdnCustomDomain.cs (hunk 7 above)
Comment (CamSoper) [inline src/ResourceManager/Cdn/Commands.Cdn/CustomDomain/RemoveAzureRmCdnCustomDomain.cs]: Then how does one skip the interactive confirmation?

File: src/ResourceManager/Cdn/Commands.Cdn/Endpoint/RemoveAzureRmCdnEndnpoint.cs [hunk 14]
```diff
@@ -54,35 +58,28 @@ public override void ExecuteCmdlet()
...
-                CdnManagementClient.Endpoints.Get(EndpointName, ProfileName, ResourceGroupName);
+                throw new PSArgumentException(string.Format(
+                    Resources.Error_DeleteNonExistingEndpoint,
+                    EndpointName,
+                    ProfileName,
+                    ResourceGroupName));
             }
-            catch (AggregateException exception)
+           
+            if(Force || ShouldProcess(string.Format(Resources.Confirm_RemoveEndpoint, EndpointName, ProfileName, ResourceGroupName)))
```
Comment (markcowl) [inline src/ResourceManager/Cdn/Commands.Cdn/Endpoint/RemoveAzureRmCdnEndnpoint.cs]: No, again, please remove force and use only ShouldProcess here

File: src/ResourceManager/Cdn/Commands.Cdn/Endpoint/SetAzureRmCdnEndpoint.cs [hunk 15]
```diff
@@ -27,7 +27,7 @@ namespace Microsoft.Azure.Commands.Cdn.Endpoint
     [Cmdlet(VerbsCommon.Set, "AzureRmCdnEndpoint"), OutputType(typeof(PSEndpoint))]
     public class SetAzureRmCdnEndpoint : AzureCdnCmdletBase
```
Comment (markcowl) [inline src/ResourceManager/Cdn/Commands.Cdn/Endpoint/SetAzureRmCdnEndpoint.cs]: This should really SupportShouldProcess and have execution guarded by ShouldProcess as well

File: src/ResourceManager/Cdn/Commands.Cdn/Endpoint/StartAzureRmCdnEndpoint.cs [hunk 16]
```diff
@@ -23,23 +23,23 @@ namespace Microsoft.Azure.Commands.Cdn.Endpoint
     [Cmdlet(VerbsLifecycle.Start, "AzureRmCdnEndpoint"), OutputType(typeof(bool))]
```
Comment (markcowl) [inline src/ResourceManager/Cdn/Commands.Cdn/Endpoint/StartAzureRmCdnEndpoint.cs]: SupportsShouldProcess=true, Guard execution by ShouldProcess

File: src/ResourceManager/Cdn/Commands.Cdn/Origin/SetAzureRmCdnOrigin.cs [hunk 17]
```diff
@@ -25,7 +25,7 @@ namespace Microsoft.Azure.Commands.Cdn.Origin
     [Cmdlet(VerbsCommon.Set, "AzureRmCdnOrigin"), OutputType(typeof(PSOrigin))]
```
Comment (markcowl) [inline src/ResourceManager/Cdn/Commands.Cdn/Origin/SetAzureRmCdnOrigin.cs]: SuppoprtsShouldProcess=true, Guard execution with SHouldProcess

File: src/ResourceManager/Cdn/Commands.Cdn/Profile/GetAzureRmCdnProfile.cs [hunk 18]
```diff
@@ -19,26 +19,29 @@
...
-        [Parameter(Mandatory = false, HelpMessage = "Azure Cdn profile name.")]
+        [Parameter(Mandatory = false, HelpMessage = "Azure CDN profile name.")]
         [ValidateNotNullOrEmpty]
         public string ProfileName { get; set; }
 
         /// <summary>
         /// The resource group name of the profile.
         /// </summary>
-        [Parameter(Mandatory = false, HelpMessage = "The resource group of the Azure Cdn Profile")]
+        [Parameter(Mandatory = false, HelpMessage = "The resource group of the Azure CDN profile.")]
```
Comment (markcowl) [inline src/ResourceManager/Cdn/Commands.Cdn/Profile/GetAzureRmCdnProfile.cs]: ValueFromPipelineByPropertyName allows piping from generic cmdlets.  Also, please add alias to 'ResourceGroup'

File: src/ResourceManager/Cdn/Commands.Cdn.Test/ScenarioTests/CustomDomainTests.ps1 (hunk 6 above)
Comment (CamSoper) [inline src/ResourceManager/Cdn/Commands.Cdn.Test/ScenarioTests/CustomDomainTests.ps1]: I get what you're saying, and believe me, I wanted to do this differently too, however the two CustomDomain tests rely on a non-injectable outside resource - a DNS CNAME that must be setup manually on DNS.  That's why the original authors ( @kuangweizhang ?  Qinyuan? ) did it this way.

File: src/ResourceManager/Cdn/Commands.Cdn/Profile/NewAzureRmCdnProfile.cs [hunk 19]
```diff
@@ -36,73 +37,61 @@ public class NewAzureRmCdnProfile : AzureCdnCmdletBase
         /// <summary>
         /// Gets or sets the profile name.
         /// </summary>
-        [Parameter(Mandatory = true, HelpMessage = "Azure Cdn profile name.")]
+        [Parameter(Mandatory = true, HelpMessage = "Azure CDN profile name.")]
```
Comment (markcowl) [inline src/ResourceManager/Cdn/Commands.Cdn/Profile/NewAzureRmCdnProfile.cs]: Add SUpportsSHouldProcess=true and Guard execution with SHouldProcess

File: src/ResourceManager/Cdn/Commands.Cdn/Profile/RemoveAzureRmCdnProfile.cs [hunk 20]
```diff
@@ -20,27 +20,32 @@
...
-        [Parameter(Mandatory = true, ParameterSetName = ObjectParameterSet, ValueFromPipeline = true, HelpMessage = "The profile.")]
+        [Parameter(Mandatory = true, ParameterSetName = ObjectParameterSet, ValueFromPipeline = true, HelpMessage = "The Azure CDN profile.")]
         [ValidateNotNullOrEmpty]
         public PSProfile CdnProfile { get; set; }
 
-        [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "Return object if specified.")]
+        [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "Return object (if specified).")]
         public SwitchParameter PassThru { get; set; }
 
+        [Parameter()]
```
Comment (markcowl) [inline src/ResourceManager/Cdn/Commands.Cdn/Profile/RemoveAzureRmCdnProfile.cs]: See previous comments on Force/SHouldProcess

File: src/ResourceManager/Cdn/Commands.Cdn/Profile/SetAzureRmCdnProfile.cs [hunk 21]
```diff
@@ -30,7 +30,7 @@ public class SetAzureRmCdnProfile : AzureCdnCmdletBase
         /// <summary>
         /// Gets or sets the profile to update.
         /// </summary>
```
Comment (markcowl) [inline src/ResourceManager/Cdn/Commands.Cdn/Profile/SetAzureRmCdnProfile.cs]: SupportsShouldProcess=true, Guard execution with SHouldProcess

**Comment by markcowl** [comment]
@CamSoper Some comments, three things that apply across the board:

(1) Your remove cmdlets *must not* use the pattern Force || ShouldProcess, this prevents ConfirmPreference and WhatIf parameters from working correctly.  The best option would be to remove the ConfirmImpact.High and simply use ShouldProcess to determine whether a prompt occurs.  This doesn't show a prompt unless the user asks for it bu lowering ConfirmPreference or specifying -Confirm.  A less good option would be to remove ConfirmImpact.High and use 
```
If (ShouldProcess(targe, message)) {
   if (Force || ShouldContinue(shouldContinuePrompt)) {
   ... cmdlet execution
  }
}
```
This acts as above, except that, in addition, the ShouldContinue prompt always occurs unless the user specifies - Force.
An even less good option would be to use only ShouldProcess with ConfirmImpact.High.  This requires users to specify ```-Conforim:$false``` to avoid prompting.  You can see: https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d for a more detailed dscussion.

(2) For any cmdlet where you have added a parameter set, the DefaultParameterSetName property og the Cmdlet attribute should be set to the old parameter set, this avoids breaking changes (and you should always have a default parameter set if there is more thna one)

(3) For any cmdlet you ar5e changing that makes a change or has side effects, I have asked for supporting the SHouldProcess pattern - this allows users to execute the cmdlet with -WhatIf to determine what it would do if executed.

**Comment by markcowl** [comment]
on demand run here:http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1058/

LGTM once the tests pass.
